# 📁 Odmena Nigeria Project
## 📄 A.I for Energy: Transitioning Toward a Sustainable Energy System
### 📃 Cleaning the Nigeria GPS CSV Dataset downloaded from the DHS Library

we will be cleaning the gps data downloaded from the DHS Dataset library.

### 📝 Info regarding accessing the Dataset

- The dataset was downloaded from the DHS Archives using this [link](https://dhsprogram.com/data/dataset_admin/login_main.cfm;jsessionid=D2884DD613D78701[…]&CFTOKEN=43b13061df228c76-8B032473-D864-1CE0-3AEA5EE27005CC99)
- You would need to create an account with DHS to access the dataset
- You would need to fill a form telling DHS what you need the dataset for and how the dataset can help you achieve your goals.
- After submitting the form, wait for about 5 days to recieve a response email from DHS. If granted permission, this email will inform you that you now have access to the dataset and also nstructions on how to access it as well as terms of use.
- Once you get a confirmation mail you can download only the dataset that you have been granted permission to access.

## 📝 Summary of Conclusion

The DHS GPS Data was downloaded from the DHS data archives. It was successfully cleaned, all outliers were removed and the cleaned data was added to the project archives

### 📝 Data Cleaning Process

#### ✔ Step 1: Import the required libraries

As in any data cleaning project. The first thing we do is import the libraries we will use. In our case it's just the `Pandas` library.

In [ ]:
# import pandas,
import pandas as pd

Since we are using colab here, we'll also need to mount our gooogle drive to the colab notebook to fetch our data.

**Note:** 
- The data should be uploaded to your google drive storage. You can do this before or after mounting.
- If you are running this notebook locally then you can skip this step

In [ ]:
from google.colab import drive

drive.mount('/content/gdrive')

Mounted at /content/gdrive


#### ✔ Step 2: Read the data into the notebook

Once the drive has been mounted, you can read in the file the same way you would on a local machine. You just need to replace `C://` with `/content/gdrive/My Drive/` as is done in the code cell below.

In [ ]:
file = '/content/gdrive/My Drive/NGGC7BFL.csv'

df = pd.read_csv(file)
df.head()

DHSID GPS_Dataset  ... Wet_Days_2010  Wet_Days_2015
0  NG201800000001    NGGE7AFL  ...      7.083333       6.446666
1  NG201800000002    NGGE7AFL  ...      7.083333       6.446666
2  NG201800000003    NGGE7AFL  ...      8.760000       8.199166
3  NG201800000004    NGGE7AFL  ...      8.410000       7.852499
4  NG201800000005    NGGE7AFL  ...      9.687500       9.210000

[5 rows x 130 columns]

#### ✔ Step 3: Drop all unwanted columns

There are a lot of columns in this dataset, a lot of them are not needed for our project so we need to drop them. The columns we drop for this cleaning are:
 
 * DHSID
 * GPS_DATASET
 * DHSCC
 * DHSYEAR 
 * DHSCLUST
 * SurveyID
 * Drought_Episodes
 * Enhanced_Vegitation_Index_1985
 * Enhanced_Vegitation_Index_1990
 * Enhanced_Vegitation_Index_1995
 * Enhanced_Vegitation_Index_2000
 * Enhanced_Vegitation_Index_2005
 * Enhanced_Vegitation_Index_2010
 * Enhanced_Vegitation_Index_2015
 * Frost_Days_2000
 * Frost_Days_2005
 * Frost_Days_2010
 * Frost_Days_2015
 * Gross_Cell_Production
 * Growing_Season_Length
 * Irrigation
 * ITN_Coverage_2005
 * ITN_Coverage_2010
 * ITN_Coverage_2015
 * Land_Surface_Temperature_2000
 * Land_Surface_Temperature_2005
 * Land_Surface_Temperature_2010
 * Livestock_Cattle
 * Livestock_Chickens
 * Livestock_Ducks
 * Livestock_Goats
 * Livestock_Pigs
 * Livestock_Sheep
 * Malaria_Incidence_2000
 * Malaria_Incidence_2005
 * Malaria_Incidence_2010
 * Malaria_Incidence_2015
 * Malaria_Prevalence_2000
 * Malaria_Prevalence_2005
 * Malaria_Prevalence_2010
 * Malaria_Prevalence_2015
 * Maximum_Temperature_2000
 * Maximum_Temperature_2005
 * Maximum_Temperature_2010
 * Mean_Temperature_2000
 * Mean_Temperature_2005
 * Mean_Temperature_2010
 * Minimum_Temperature_2000
 * Minimum_Temperature_2005
 * Minimum_Temperature_2010
 * Night_Land_Surface_Temp_2000
 * Night_Land_Surface_Temp_2005
 * Night_Land_Surface_Temp_2010
 * PET_2000
 * PET_2005
 * PET_2010
 * PET_2015
 * Proximity_to_National_Borders
 * Proximity_to_Protected_Areas
 * Rainfall_1985
 * Rainfall_1990
 * Rainfall_1995
 * Rainfall_2000
 * Rainfall_2005
 * Rainfall_2010
 * Slope
 * SMOD_Population_1990
 * SMOD_Population_2000
 * SMOD_Population_2015
 * Temperature_April
 * Temperature_August
 * Temperature_December
 * Temperature_February
 * Temperature_January
 * Temperature_July
 * Temperature_June
 * Temperature_March
 * Temperature_May
 * Temperature_November
 * Temperature_October
 * Temperature_September
 * Travel_Times_2000
 * Travel_Times_2015
 * U5_Population_2000
 * U5_Population_2005
 * U5_Population_2010
 * U5_Population_2015
 * UN_Population_Count_2000
 * UN_Population_Count_2005
 * UN_Population_Count_2010
 * UN_Population_Density_2000
 * UN_Population_Density_2005
 * UN_Population_Density_2010
 * UN_Population_Density_2015
 * Wet_Days_2000
 * Wet_Days_2005
 * Wet_Days_2010


 A list of all columns before dropping is shown below:

In [ ]:
df.columns

Index(['DHSID', 'GPS_Dataset', 'DHSCC', 'DHSYEAR', 'DHSCLUST', 'SurveyID',
       'All_Population_Count_2005', 'All_Population_Count_2010',
       'All_Population_Count_2015', 'Annual_Precipitation_2000',
       ...
       'UN_Population_Count_2010', 'UN_Population_Count_2015',
       'UN_Population_Density_2000', 'UN_Population_Density_2005',
       'UN_Population_Density_2010', 'UN_Population_Density_2015',
       'Wet_Days_2000', 'Wet_Days_2005', 'Wet_Days_2010', 'Wet_Days_2015'],
      dtype='object', length=130)

In [ ]:
column = df.columns

for col in column:
  print(col)

DHSID
GPS_Dataset
DHSCC
DHSYEAR
DHSCLUST
SurveyID
All_Population_Count_2005
All_Population_Count_2010
All_Population_Count_2015
Annual_Precipitation_2000
Annual_Precipitation_2005
Annual_Precipitation_2010
Annual_Precipitation_2015
Aridity_2000
Aridity_2005
Aridity_2010
Aridity_2015
BUILT_Population_1990
BUILT_Population_2000
BUILT_Population_2014
Day_Land_Surface_Temp_2000
Day_Land_Surface_Temp_2005
Day_Land_Surface_Temp_2010
Day_Land_Surface_Temp_2015
Diurnal_Temperature_Range_2000
Diurnal_Temperature_Range_2005
Diurnal_Temperature_Range_2010
Diurnal_Temperature_Range_2015
Drought_Episodes
Enhanced_Vegetation_Index_1985
Enhanced_Vegetation_Index_1990
Enhanced_Vegetation_Index_1995
Enhanced_Vegetation_Index_2000
Enhanced_Vegetation_Index_2005
Enhanced_Vegetation_Index_2010
Enhanced_Vegetation_Index_2015
Frost_Days_2000
Frost_Days_2005
Frost_Days_2010
Frost_Days_2015
Global_Human_Footprint
Gross_Cell_Production
Growing_Season_Length
Irrigation
ITN_Coverage_2005
ITN_Coverage_2010
ITN_Co

In [ ]:
column

Index(['DHSID', 'GPS_Dataset', 'DHSCC', 'DHSYEAR', 'DHSCLUST', 'SurveyID',
       'All_Population_Count_2005', 'All_Population_Count_2010',
       'All_Population_Count_2015', 'Annual_Precipitation_2000',
       ...
       'UN_Population_Count_2010', 'UN_Population_Count_2015',
       'UN_Population_Density_2000', 'UN_Population_Density_2005',
       'UN_Population_Density_2010', 'UN_Population_Density_2015',
       'Wet_Days_2000', 'Wet_Days_2005', 'Wet_Days_2010', 'Wet_Days_2015'],
      dtype='object', length=130)

In [ ]:
df1 = df[['All_Population_Count_2015','Annual_Precipitation_2015','Aridity_2015',
          'BUILT_Population_2014','Day_Land_Surface_Temp_2015','Diurnal_Temperature_Range_2015',
          'Global_Human_Footprint','Land_Surface_Temperature_2015','Maximum_Temperature_2015',
          'Mean_Temperature_2015','Minimum_Temperature_2015','Nightlights_Composite',
          'Night_Land_Surface_Temp_2015','Proximity_to_Water','Rainfall_2015','UN_Population_Count_2015',
          'Wet_Days_2015']]

In [ ]:
df1.head()

All_Population_Count_2015  ...  Wet_Days_2015
0               94469.648438  ...       6.446666
1               61496.667969  ...       6.446666
2               57563.781250  ...       8.199166
3               28072.353516  ...       7.852499
4                7109.328613  ...       9.210000

[5 rows x 17 columns]

#### ✔ Step 4: Look for missing values

Now we use the `df.info()` command to see the count of all non-null rows in each column. This will help use see if we have any null values in our dataset.

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1389 entries, 0 to 1388
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   All_Population_Count_2015       1389 non-null   float64
 1   Annual_Precipitation_2015       1389 non-null   float64
 2   Aridity_2015                    1389 non-null   float64
 3   BUILT_Population_2014           1389 non-null   float64
 4   Day_Land_Surface_Temp_2015      1389 non-null   float64
 5   Diurnal_Temperature_Range_2015  1389 non-null   float64
 6   Global_Human_Footprint          1389 non-null   float64
 7   Land_Surface_Temperature_2015   1389 non-null   float64
 8   Maximum_Temperature_2015        1389 non-null   float64
 9   Mean_Temperature_2015           1389 non-null   float64
 10  Minimum_Temperature_2015        1389 non-null   float64
 11  Nightlights_Composite           1389 non-null   float64
 12  Night_Land_Surface_Temp_2015    13

From the code cell above, all the columns have the same amount of non-null values (1389 of them). 
This tells us that we have no missing values in the data.

#### ✔ Step 5: Look for outliers

Now we look for outliers in our dataset. We can do this using the `df.describe()` method.

In [ ]:
df1.describe()

All_Population_Count_2015  ...  Wet_Days_2015
count               1.389000e+03  ...    1389.000000
mean                1.039164e+05  ...    -438.870507
std                 1.660669e+05  ...    2067.192648
min                -9.999000e+03  ...   -9999.000000
25%                 2.274591e+04  ...       4.749166
50%                 5.241218e+04  ...       7.750000
75%                 1.242183e+05  ...       9.780000
max                 2.482358e+06  ...      14.435833

[8 rows x 17 columns]

After running the command, we see that all columns have a minimum value of `-9999.00`. This value seems too low of our data. For example, population counts cannot be a negative value and temperature values should not go as low as `-9999.0`.

We will have to deal with these. I chose to remove the rows containing these values from the dataset since we have a lot of data.

In [ ]:
df1[df1['Annual_Precipitation_2015'] == -9999.0]

All_Population_Count_2015  ...  Wet_Days_2015
47                 -9999.000000  ...        -9999.0
121                -9999.000000  ...        -9999.0
204                -9999.000000  ...        -9999.0
334                -9999.000000  ...        -9999.0
335                -9999.000000  ...        -9999.0
...                         ...  ...            ...
1234                1355.886475  ...        -9999.0
1235              148111.453125  ...        -9999.0
1286               -9999.000000  ...        -9999.0
1310               37859.808594  ...        -9999.0
1327               -9999.000000  ...        -9999.0

[62 rows x 17 columns]

In [ ]:
df1[df1['Annual_Precipitation_2015'] == -9999.0].value_counts()

All_Population_Count_2015  Annual_Precipitation_2015  Aridity_2015  BUILT_Population_2014  Day_Land_Surface_Temp_2015  Diurnal_Temperature_Range_2015  Global_Human_Footprint  Land_Surface_Temperature_2015  Maximum_Temperature_2015  Mean_Temperature_2015  Minimum_Temperature_2015  Nightlights_Composite  Night_Land_Surface_Temp_2015  Proximity_to_Water  Rainfall_2015  UN_Population_Count_2015  Wet_Days_2015
-9999.000000               -9999.0                    -9999.0       -9999.000000           -9999.000000                -9999.0                         -9999.000000            -9999.000000                   -9999.0                   -9999.0                -9999.0                   -9999.000000           -9999.000000                  -9999.000000        -9999.000000   -9999.000000              -9999.0          30
 92894.773438              -9999.0                    -9999.0        0.928832               29.388000                  -9999.0                          56.642857               

In [ ]:
df1 = df1[df1['Annual_Precipitation_2015'] != -9999.0]

df1.head()

All_Population_Count_2015  ...  Wet_Days_2015
0               94469.648438  ...       6.446666
1               61496.667969  ...       6.446666
2               57563.781250  ...       8.199166
3               28072.353516  ...       7.852499
4                7109.328613  ...       9.210000

[5 rows x 17 columns]

In [ ]:
df1.describe()

All_Population_Count_2015  ...  Wet_Days_2015
count               1.327000e+03  ...    1327.000000
mean                1.043153e+05  ...       7.797186
std                 1.655601e+05  ...       2.883874
min                 2.335047e+02  ...       2.309167
25%                 2.414434e+04  ...       5.165000
50%                 5.357721e+04  ...       8.129167
75%                 1.241933e+05  ...       9.962500
max                 2.482358e+06  ...      14.435833

[8 rows x 17 columns]

In [ ]:
df1[df1['Day_Land_Surface_Temp_2015'] == -9999.0]

All_Population_Count_2015  ...  Wet_Days_2015
1128               23855.457031  ...      12.961666

[1 rows x 17 columns]

In [ ]:
df1 = df1.drop(1128, axis=0)

df1

All_Population_Count_2015  ...  Wet_Days_2015
0                  94469.648438  ...       6.446666
1                  61496.667969  ...       6.446666
2                  57563.781250  ...       8.199166
3                  28072.353516  ...       7.852499
4                   7109.328613  ...       9.210000
...                         ...  ...            ...
1384               48227.656250  ...       8.818334
1385              191277.156250  ...       8.147500
1386               19350.255859  ...       8.572500
1387               74482.679688  ...       8.983334
1388              389235.968750  ...       8.818334

[1326 rows x 17 columns]

In [ ]:
df1.describe()

All_Population_Count_2015  ...  Wet_Days_2015
count               1.326000e+03  ...    1326.000000
mean                1.043760e+05  ...       7.793292
std                 1.656078e+05  ...       2.881468
min                 2.335047e+02  ...       2.309167
25%                 2.420504e+04  ...       5.156666
50%                 5.373700e+04  ...       8.105416
75%                 1.242058e+05  ...       9.962500
max                 2.482358e+06  ...      14.435833

[8 rows x 17 columns]

In [ ]:
df1[df1['Rainfall_2015'] == -9999.0]

All_Population_Count_2015  ...  Wet_Days_2015
1121               67134.265625  ...      12.124166
1122              102461.734375  ...      12.124166

[2 rows x 17 columns]

In [ ]:
df1 = df1.drop([1121, 1122], axis= 0)

df1.head()

All_Population_Count_2015  ...  Wet_Days_2015
0               94469.648438  ...       6.446666
1               61496.667969  ...       6.446666
2               57563.781250  ...       8.199166
3               28072.353516  ...       7.852499
4                7109.328613  ...       9.210000

[5 rows x 17 columns]

Now that we have finished removing all outliers, we should take a look at the rest of the data to make sure it is properly cleaned. 

In [ ]:
df1.describe()

All_Population_Count_2015  ...  Wet_Days_2015
count               1.324000e+03  ...    1324.000000
mean                1.044056e+05  ...       7.786750
std                 1.657297e+05  ...       2.878718
min                 2.335047e+02  ...       2.309167
25%                 2.415551e+04  ...       5.148333
50%                 5.356173e+04  ...       8.081250
75%                 1.243021e+05  ...       9.962500
max                 2.482358e+06  ...      14.435833

[8 rows x 17 columns]

#### ✔ Step 6: Save our cleaned data into  file

Now that our data is properly cleaned. We should save our data in a file. I chose to save it as a csv file.

We will achieve this using the `df.to_csv()` method from `pandas`.

In [ ]:
df1.to_csv('/content/gdrive/My Drive/NGGC7BFL-cleaned.csv', index=False)

## 🏁 Conclusion

The DHS GPS Data was successfully cleaned, all outliers were removed and the cleaned data was added to the project archives